compiler/rmi/rmi-interface/src/org/jetbrains/kotlin/rmi/DaemonParams.kt

```diff
@@ -0,0 +1,272 @@
+                else false } }
+            if (parser != null) {
+                val optionLength = matchingOption.length()
+                when {
+                    parser is BoolPropParser<*,*> ->
+                        if (param.length() > optionLength) throw IllegalArgumentException("Invalid switch option '$param', expecting $matchingOption without arguments")
+                        else parser.apply("")
+                    param.length() > optionLength ->
+                        if (param[optionLength] != '=') {
+                            if (parser.allowMergedArg) parser.apply(param.substring(optionLength))
+                            else throw IllegalArgumentException("Invalid option syntax '$param', expecting $matchingOption[= ]<arg>")
+                        }
+                        else parser.apply(param.substring(optionLength + 1))
+                    else -> currentParser = parser
+                }
+                false
+            }
+            else if (restParser != null && param.startsWith(prefix)) {
+                restParser.add(param.removePrefix(prefix))
+                false
+            }
+            else true
+        }
+        else {
+            currentParser!!.apply(param)
+            currentParser = null
+            false
+        }
+    }
+    if (currentParser != null) throw IllegalArgumentException("Expecting argument for the option $matchingOption")
+    return res
+}
+
+// TODO: find out how to create more generic variant using first constructor
+//fun<C> C.propsToParams() {
+//    val kc = C::class
+//    kc.constructors.first().
+//}
+
+
+
+public interface CmdlineParams : Serializable {
+    public val extractors: List<PropExtractor<*,*,*>>
+    public val parsers: List<PropParser<*,*,*>>
+}
+
+public fun Iterable<String>.filterSetProps(vararg cs: CmdlineParams, prefix: String) : Iterable<String> =
+    filterSetProps(cs.flatMap { it.parsers }, prefix)
+
+
+public data class DaemonLaunchingOptions(
+        public var maxMemory: String = "",
+        public var maxPermSize: String = "",
+        public var reservedCodeCacheSize: String = "",
+        public var otherJvmParams: MutableCollection<String> = arrayListOf()
+) : CmdlineParams {
+
+    override val extractors: List<PropExtractor<*,*,*>>
+        get() = listOf( PropExtractor(this, ::maxMemory, "Xmx", skipIf = { it.isEmpty() }, mergeWithDelimiter = ""),
+                        PropExtractor(this, ::maxPermSize, "XX:MaxPermSize", skipIf = { it.isEmpty() }, mergeWithDelimiter = "="),
+                        PropExtractor(this, ::reservedCodeCacheSize, "XX:ReservedCodeCacheSize", skipIf = { it.isEmpty() }, mergeWithDelimiter = "="),
+                        RestPropExtractor(this, ::otherJvmParams))
+
+    override val parsers: List<PropParser<*,*,*>>
+        get() = listOf( PropParser(this, ::maxMemory, listOf("Xmx"), { it }, allowMergedArg = true),
```

Comment:
i'd extract `{ it }` or use `identity()` from `core/util.runtime/src/org/jetbrains/kotlin/utils/functions.kt`